File: rust-simpletron/src/simpletron/assembler/lexer.rs
use crate::simpletron::utils::Character;

#[derive(Clone, Debug)]
#[allow(dead_code)]
pub enum Token{ Name(String), Number(String), Colon,  None}

pub struct Lexer{ index:usize, current:Character, data: String, to_newline: bool }
impl Lexer{
    pub fn new(data:&str)->Self{
        let string = String::from(data);
        let init = Character::new(string.chars().nth(0).unwrap());
        Lexer{ index:0, current:init, data: string, to_newline: false }
    }
    
    pub fn has_next(&mut self)->bool{
        while self.index < self.data.len(){
            self.current = Character::new(self.data.chars().nth(self.index).unwrap());
			if self.to_newline && self.current.unwrap() == '\n'{
				self.to_newline = false;
			}else if !self.to_newline && !self.current.is_whitespace(){
				return true; 
			} 
            self.index += 1;
        }
        return false;
    }
    
    fn pop(&mut self)->char{
        let init = self.data.chars().nth(self.index).unwrap();
        self.index += 1;
        return init;
    }
    
    pub fn get_next_token(&mut self)->Result<Token, String>{
		if self.current.is_alphabetic(){
			return self.get_name_token();
		}
		
		if self.current.is_numeric(){
			return self.get_number_token();
		}
		match self.current.unwrap(){
			':' => { self.pop(); return Ok(Token::Colon); },
			__  => return Result::Err(format!("unexpected token {} encountered", self.pop()))
		}
	}

	fn get_name_token(&mut self)->Result<Token, String>{
		let mut builder = String::new();
		while self.index < self.data.len(){
            self.current = Character::new(self.data.chars().nth(self.index).unwrap());
            let passable = !self.current.is_alphanumeric();
            if passable { break; }else { builder.push(self.current.unwrap()); }
            self.index += 1;
        }
		return Ok(Token::Name(builder));
	}

	fn get_number_token(&mut self)->Result<Token, String>{
		let mut builder = String::new();
		while self.index < self.data.len(){
            self.current = Character::new(self.data.chars().nth(self.index).unwrap());
            let important = !self.current.is_whitespace() && self.current.is_hexdigit();
            if important { builder.push(self.current.unwrap()); } else { break; };
            self.index += 1;
        }
		return Ok(Token::Number(builder));
	}
}

File: cpp-simpletron/src/lexer.cpp
#include "headers/lexer.h"

using namespace simpletron::assembler;
using namespace simpletron::utils;

Lexer::Lexer(std::string data){
    this->data = data;
}

char Lexer::pop(){
    char init = this->data[this->index];
    this->index += 1;
    return init;
}

bool Lexer::hasNext(){
    while(this->index < this->data.length()){
        this->current = new utils::Character(this->data[this->index]);
        if(this->to_newline && this->current->unwrap() == '\n'){
            this->to_newline = false;
        }else if(!this->to_newline && !this->current->isWhitespace()){
            return true; 
        }
        this->index += 1;
    }
    return false;
}

Result<Token>* Lexer::getNameToken(){
    std::string builder = "";
    while(this->index < this->data.length()){
        this->current = new Character(this->data[this->index]);
        bool passable = !this->current->isAlphanumeric();
        if(passable){
            break;
        }else {
            builder += this->current->unwrap();
        }
        this->index += 1;
    }
    return Result<Token>::Ok(new Token(TokenType::Name, builder));
}

Result<Token>* Lexer::getNumberToken(){
    std::string builder = "";
    while(this->index < this->data.length()){
        this->current = new Character(this->data[this->index]);
        bool important = !this->current->isWhitespace() && this->current->isHexdigit();
        if(important){
            builder += this->current->unwrap();
        }else{
            break;
        };
        this->index += 1;
    }
    return Result<Token>::Ok(new Token(TokenType::Number, builder));
}

Result<Token>* Lexer::nextToken(){
    if(this->current->isAlphabetic()){
        return this->getNameToken();
    }else if(this->current->isNumeric()){
        return this->getNumberToken();
    }else if(this->current->unwrap() == ':'){
        this->pop();
        return Result<Token>::Ok(new Token(TokenType::Colon, ":"));
    }
    std::string s(1, this->pop());
    return Result<Token>::Error("unexpected token {" + s + "} encountered");
}